_posts/2020-07-31-fnn-vae-for-noisy-timeseries/fnn_vae_for_noisy_timeseries.Rmd

```diff
@@ -0,0 +1,509 @@
+  train_kl$reset_states()
+  
+}))
+
+```
+
+To finish up the model section, here is the actual training code. This is nearly identical to what we did for FNN-LSTM before.
+
+```{r, eval = FALSE}
+n_latent <- 10L
+n_features <- 1
+
+encoder <- vae_encoder_model(n_timesteps,
+                         n_features,
+                         n_latent)
+
+decoder <- vae_decoder_model(n_timesteps,
+                         n_features,
+                         n_latent)
+mse_loss <-
+  tf$keras$losses$MeanSquaredError(reduction = tf$keras$losses$Reduction$SUM)
+
+train_loss <- tf$keras$metrics$Mean(name = 'train_loss')
+train_fnn <- tf$keras$metrics$Mean(name = 'train_fnn')
+train_mse <-  tf$keras$metrics$Mean(name = 'train_mse')
+train_kl <-  tf$keras$metrics$Mean(name = 'train_kl')
+
+fnn_multiplier <- 1 # default value used in nearly all cases (see text)
+fnn_weight <- fnn_multiplier * nrow(x_train)/batch_size
+
+kl_weight <- 1
+
+optimizer <- optimizer_adam(lr = 1e-3)
+
+for (epoch in 1:100) {
+  cat("Epoch: ", epoch, " -----------\n")
+  training_loop_vae(ds_train)
+ 
+  test_batch <- as_iterator(ds_test) %>% iter_next()
+  encoded <- encoder(test_batch[[1]][1:1000])
+  test_var <- tf$math$reduce_variance(encoded, axis = 0L)
+  print(test_var %>% as.numeric() %>% round(5))
+}
+```
+
+## Experimental setup and data
+
+The idea was to add white noise to a deterministic series. This time, the [Roessler
+system](https://en.wikipedia.org/wiki/R%C3%B6ssler_attractor) was chosen, mainly for the prettiness of its attractor, apparent
+even in its two-dimensional projections:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Roessler attractor, two-dimensional projections."}
+knitr::include_graphics("images/roessler.png")
+```
+
+Like we did for the Lorenz system in the first part of this series, we use `deSolve` to generate data from the Roessler
+equations.
+
+```{r, eval = FALSE}
+library(deSolve)
+
+parameters <- c(a = .2,
+                b = .2,
+                c = 5.7)
+
+initial_state <-
+  c(x = 1,
+    y = 1,
+    z = 1.05)
+
+roessler <- function(t, state, parameters) {
+  with(as.list(c(state, parameters)), {
+    dx <- -y - z
+    dy <- x + a * y
+    dz = b + z * (x - c)
+    
+    list(c(dx, dy, dz))
+  })
+}
+
+times <- seq(0, 2500, length.out = 20000)
+
+roessler_ts <-
+  ode(
+    y = initial_state,
+    times = times,
+    func = roessler,
+    parms = parameters,
+    method = "lsoda"
+  ) %>% unclass() %>% as_tibble()
+
+n <- 10000
+roessler <- roessler_ts$x[1:n]
+
+roessler <- scale(roessler)
+```
+
+Then, noise is added, to the desired degree, by drawing from a normal distribution, centered at zero, with standard deviations
+varying between 1 and 2.5.
+
+```{r, eval = FALSE}
+# add noise
+noise <- 1 # also used 1.5, 2, 2.5
+roessler <- roessler + rnorm(10000, mean = 0, sd = noise)
+
+```
+
+Here you can compare effects of adding no (left), standard deviation-1 (middle), and standard deviation-2.5 Gaussian noise:
```

Comment:
```suggestion
Here you can compare effects of adding no standard deviation (left), standard deviation-1 (middle), and standard deviation-2.5 Gaussian noise:
```

Reply:
Or you could change "no" to "nothing"

Reply:
hm, that was supposed to refer to "noise" an infinity of words later ;-)
could replace by "no noise" I guess